diff --git a/walk/walk.go b/walk/walk.go
index 8c9859a..410c64c 100644
--- a/walk/walk.go
+++ b/walk/walk.go
@@ -385,6 +385,12 @@ func newWalker(c *config.Config, cexts []config.Configurer, dirs []string, mode
 	}
 	// END-PATCH(gitignore) ---
 
+	// PATCH(watch) ---
+	if hook, ok := c.Exts["aspect:walkCache:load"].(func(any)); ok {
+		hook(&w.cache.entryMap)
+	}
+	// END-PATCH(watch) ---
+
 	// Asynchronously populate the walker cache in the background.
 	go w.populateCache(mode)
 
