diff --git a/toolchain/cc_toolchain_config.bzl b/toolchain/cc_toolchain_config.bzl
--- a/toolchain/cc_toolchain_config.bzl
+++ b/toolchain/cc_toolchain_config.bzl
@@ -365,6 +368,6 @@ def cc_toolchain_config(
         "gcc": wrapper_bin_prefix + "cc_wrapper.sh",
         "gcov": tools_path_prefix + "llvm-profdata",
-        "ld": tools_path_prefix + "ld.lld",
+        "ld": tools_path_prefix + ("ld64.lld" if target_os == "darwin" else "ld.lld"),
         "llvm-cov": tools_path_prefix + "llvm-cov",
         "llvm-profdata": tools_path_prefix + "llvm-profdata",
         "nm": tools_path_prefix + "llvm-nm",
diff --git a/toolchain/internal/common.bzl b/toolchain/internal/common.bzl
--- a/toolchain/internal/common.bzl
+++ b/toolchain/internal/common.bzl
@@ -45,2 +45,3 @@ _toolchain_tools_darwin = {
     "llvm-libtool-darwin": "libtool",
+    "ld64.lld": "ld64.lld",
 }
